### DISCUSSION TITLE
Command Combos

### DISCUSSION THREAD

**Discussion by madprops** [madprops/grasshopper#130] [General]
Next version will have this:

You can define 10 combos. Each combo is a list of commands:

![combos1](https://github.com/user-attachments/assets/72910e85-6285-4279-88cc-b24e631d82da)

![combos2](https://github.com/user-attachments/assets/b45d8004-ac9c-44ab-880e-607c036ae828)

They appear in the palette:

![combos23](https://github.com/user-attachments/assets/e57225ad-f8cd-44bc-b156-7b66b527da69)

And they can be assigned like normal commands:

![combos34](https://github.com/user-attachments/assets/b2a72f30-8c35-49ea-82d6-c9c1898780a8)

**Comment by N3C2L** [comment]
Wow! Love that!
I just tried it, but have a problem:
I have a combo here with this 3 cmds:
Filter Domain
Close Visible
Enter Key
But the problem is, that the enter key is not working for the popoup that will close visible tabs.
Can you confirm that?
Edit: Just realized that I would need another double enter for the close visible popup.
And I have another combo:
Unload others
Enter Key
But I can't add another Enter Key in the combos menus (2x Enter key after another), that would be cool to confirm the double enter in the unload others popup.

**Comment by madprops** [comment]
I'll look into it. It's a primary key issue since 'cmd' is the primary key so it doesn't allow duplicates. I dealt with this a few days ago on a similar issue, i'll look into it

Other thing to keep in mind is that they're executed top to bottom, so you might have to order the list.

Maybe I should make it so adding items are appended instead of prepended.

**Comment by N3C2L** [comment]
Thanks for looking into it! From top to bottom, yes.
It's a good idea to make new items appended! Then we don't have to reorder them manually afterwards.

**Comment by madprops** [comment]
Ok next version will fix this. Turns out you don't need a second Enter. The synthetic Enter was triggering a `keydown` but not a `keyup` and I check enter keys on `keyup`. So now I'm making all Keys to trigger both keydown and keyup. (Well you need 2 enters to confirm the popup, this won't work if 0 items are selected)

Also combos will be able to add duplicate commands since I fixed the primary key issue.

And now list items are appended.

These new settings might be useful:

![autoclsoe](https://github.com/user-attachments/assets/ada731e1-8690-4c2f-8b9a-72f6f7f2ee95)

These will be in `More`.

Also this one:

![apend34](https://github.com/user-attachments/assets/29799aa8-bf09-4a7d-ac04-87acbc8f626e)

**Comment by N3C2L** [comment]
Thank you! All of your changes work flawlessly in my combos. It's truly a pleasure to bite into these combo burgers. And thank you also for the new settings "tab close pins" and "tab close unloaded" - I've naturally activated both of them right away.

**Comment by N3C2L** [comment]
I have another small problem 😁 
I often use edit title cmd for renaming my tabs. I'd love a combo for this, and created one:
Edit Title
Left Key
But I think the left arrow key doesn't register here in this renaming dialog.

**Comment by madprops** [comment]
Good timing, I was about to upload new version. I'm taking a look.

**Comment by madprops** [comment]
I don't think that's possible. I tried dispatching the event on the input directly too:

`
Triggering a keydown event programmatically using new KeyboardEvent('keydown', { key: 'ArrowLeft' }) might not produce the desired effect because many browser behaviors associated with keyboard events are considered "trusted" events, meaning they need to originate from actual user actions to be recognized fully.
`

Is the intended action to place the caret at the start of the title? This could be fixed through a setting

Right now it auto highlights the text if `Edit Title Auto` in More is checked

**Reply by N3C2L** [reply]
I see. Yes, you are right. It's intended to place the caret at the start of the title. I do this often on video tabs for example "min24:20" at the start of the tab, where I left the video and unload the tab afterwards.

**Comment by madprops** [comment]
Well the highlight behavior is the default since you requested it if I remember correctly. I can change the default of how this works. Fill the title but instead of highlighting put caret at start. Though highlighting might be more flexible since you can press left arrow or right arrow.

**Reply by N3C2L** [reply]
Yes I was requesting it, thanks for implementing it!
That would be great if this could be changed to put the caret at the start.

**Comment by madprops** [comment]
I'm changing that default but added a setting for highlight which can be enabled to restore old behavior.

![hh45](https://github.com/user-attachments/assets/6149a2c3-084d-4f16-9fe3-6ba7b593853a)


Actually I'm gonna add this to General:

![promptomdoe](https://github.com/user-attachments/assets/aa15947a-8655-4c2e-a28d-3f6257b2f1a6)

**Reply by N3C2L** [reply]
Great! A small problem is, when I have added the edit title cmd to a browser cmd in about:addons, then the edit title dialog gets no focus for input. You can test this when changing tabs and try this browser cmd with the keyboard. I think it's because it's focusing the website instead of the renaming dialog. It's working only when you click the tab before.

**Comment by madprops** [comment]
Yeah I don't think that's possible yet

https://bugzilla.mozilla.org/show_bug.cgi?id=1912558

**Reply by N3C2L** [reply]
Hmm, I'd be appreciate if you could look into the sidebery code, because it's working there.

**Reply by N3C2L** [reply]
Sorry I was wrong, apologize.

**Reply by N3C2L** [reply]
I was trying to run other commands before the edit title cmd with a combo, (for example pin+unpin) but that doesn't bring the input focus either.

**Reply by madprops** [reply]
If sidebery can focus the sidebar then it must be possible. But seems it can't but I haven't tried it

**Reply by N3C2L** [reply]
No I was wrong, sorry, they don't have a shortcut in the about:addons page for renaming the tab.

**Reply by madprops** [reply]
I mean mostly focusing the sidebar. I tried myself but didn't work

**Reply by N3C2L** [reply]
Oh, sorry I was thinking that when you are adding shortcuts in the about:addons page then the sidebar gets focus because shortcuts work.

**Reply by madprops** [reply]
Shortcuts work because there's signals sent from the background script to the sidebar. But the sidebar doesn't get actual focus as you need to click it manually

**Reply by N3C2L** [reply]
Alright, thanks for this clarification.

**Reply by N3C2L** [reply]
Hmm, a solution would be to run the grasshopper popup instead of the sidebar. Then the input is there. For cmds useful where keyboard input afterwards is useful:
"close menu"
"add tag"
"edit title"
Could be 5 other "popup" shortcuts in the "browser" section. When added there, the popup will run with the shortcuts set in about:addons.

Edit: There could be a setting that automatically closes the grasshopper popup if a prompt has been closed. This would be awesome for quick editing, and we don't need to press the ESC key after the prompt to escape the grasshopper popup.

**Reply by madprops** [reply]
Next version will have 10 of those.

https://github.com/user-attachments/assets/53c000f2-1568-4c61-b1c1-54f31352b278

It auto closes on prompt hide if it was spawned via a popup command.

This is going a bit into 'hack' territory as it relies on timeouts, but it works.

Also I'm gonna make Prompt Mode: At End, the default as it is more natural but you can change it in General.

**Reply by N3C2L** [reply]
This is awesome! Thanks!

**Comment by N3C2L** [comment]
It would be very convenient if a user could rename combos. 

Especially when many commands are to be executed in sequence, one wants to be absolutely sure with a proper naming of the combos before executing them. 

Also, because it would make the menus containing these combos easier to read. I think if colors can be named, it should also be possible with combos. 

What do you think about that?

**Reply by madprops** [reply]
Yeah makes sense. Also should name the new Generic Menus.

I might make it so name and icon can be defined.

**Comment by madprops** [comment]
![combos234jj](https://github.com/user-attachments/assets/069d8db4-cc7b-4e07-b21b-43a009233e50)

**Comment by madprops** [comment]
Also added a shortcut when you click Save on an addlist. If you hold Shift or Ctrl when clicking, it will reverse the insert method. For instance instead of appending to the bottom of the list, it will prepend it.

**Comment by madprops** [comment]
![genmenu34](https://github.com/user-attachments/assets/200b7028-ba28-41cb-a2b2-db31df122dcc)

**Comment by madprops** [comment]
I'm thinking now if these should be created dynamically in infinite number.

However there's no addlist within addlist ability right now.

**Comment by madprops** [comment]
An option is to create an addlist where combos can be created but with a fixed number of possible commands, like 5 commands. Do you think more than 5 commands might be needed in a combo? I would also add a delay option per combo. I like that idea because Combos are basically command factories with 1 or more commands.

**Comment by madprops** [comment]
It would look like this:

![command](https://github.com/user-attachments/assets/e24c9b7f-2ea6-404c-b2e0-f94c3f1ddb94)

**Reply by madprops** [reply]
I'll make it 10 commands.

**Comment by N3C2L** [comment]
Well, I like this idea of the infinite list of factory combos 😁 
Yeah, I think the number of max 10 is more than enough for everything.
Also very good the icon idea with every combo. Make the menus shine even more with that. Thanks for that!

**Reply by madprops** [reply]
Don't worry I think I won't change it. At least not the generic menus. Not sure what I'll do.

Well I guess we'll have to remake our command combos now.

I'm moving it to General.

But it's more orderly I guess.

**Reply by N3C2L** [reply]
I'm totally fine with 10 combos as it is now. I don't even have all 10 combos slots filled with cmds. (I have 5 filled atm.). I don't know what you mean with the generic menus.

**Reply by N3C2L** [reply]
Sorry if I misunderstood what you wrote, not a native speaker. What did you mean with "dont worry I think wont change it?"

**Reply by madprops** [reply]
I was going to revert the changes. But I think it's better this way. 

Didn't want to bother you with re-making the combos.

![command234](https://github.com/user-attachments/assets/e6f06e77-1212-4ef9-9105-cfee334d78a9)

**Comment by madprops** [comment]
I'll leave it as it was then. But with names and icons.

**Reply by N3C2L** [reply]
Ah, got it, thank you 😁

**Comment by madprops** [comment]
Ok new version up.

**Comment by madprops** [comment]
Take note of current combos because I am gonna change it to the addlist mode.

They'll be in `General` as an addlist next to Custom URLs.

**Reply by N3C2L** [reply]
Alright, I took screenshots of them. Thanks for the addlist mode! 👍

**Comment by madprops** [comment]
I'm enhancing addlist command widgets so they can be re-ordered by drag and dropping.

Also I'm making the ability to compose combos inside other combos, which wait until the combo item finishes before running the next item.

**Reply by N3C2L** [reply]
Good enhancements, thanks!

**Comment by N3C2L** [comment]
I guess there is a little problem with combos, I'd be cool if you could look into it:
Specifically the "step back" command when it comes to focusing on the last used tab: Suppose you filter the list with the "Filter Titled" command and then use the "step back" command to focus on the active tab. The problem arises that the filtering mode reverts to "All" and focuses the tab there. But what if someone wants to focus on the active tab in the "Filter Titled" filtering with a command or combo? This is currently not possible with the step back command.

Small other problem but below also a combo relation:
Maybe this is a bug: Please repro: Settings>Trigger>Click Press Item set it to "Filter Loaded".
Then try to "click press" a tab with it. Result: The actual tab is not selected in the "Filter Loaded" view. But this is strange, because if I set "Click Press" to "Unload Tab" the tab will get selected, because otherwise the tab wouldn't get unloaded. So this cmd is working there.
But due to this problem from above: Do you think it would make sense to add commands like "select tab under mouse cursor" or "left mouse click" to the available combo options? Because what if other commands or combos do need this tab activation for example keyboard shortcuts when executed from them.

**Comment by madprops** [comment]
#1 is confusing. Active means the current focused/open tab btw. So I guess you mean selected  tab, but that could be anything, not sure I see a problem here.

#2 I'm not seeing the filter load problem, when I click press a tab I see it being included in the filter. Maybe the tab is unloaded in your case?  (Oh you mean is not selected I see).

I'll check if those 2 command additions make sense.

I'm gonna add left click command but I don't know if it's gonna help.

What I'm gonna do is keep selected item on filters, through a setting, default true I guess. So you could select the item first and then do the filter loaded and it will keep the item selected.

![filtersk34](https://github.com/user-attachments/assets/48aaf90c-95d1-4eed-8133-0224df46b7f9)

**Comment by N3C2L** [comment]
1. Thank you for the new "Filter Keep Selected" function, it's very helpful for switching between filters while maintaining the current tab. However, it might have been good to introduce a command like "Focus current tab". Because when I'm within a filter, e.g., "Filter Loaded", and then scroll somewhere, the "step back" command would take me back to the "All" filter. However, with new cmd "Focus current tab" or maybe "Focus active Tab", one would want to stay in the current filter "Filter Loaded" and focus the active tab there instead of using the "Step Back" cmd wihich takes him to the "All" view and focuses the tab here.

2. Thank you for the new "Left Click" command. With this, I've now been able to create a combo on "Click Press" that goes: 1. Left Click 2. Filter Loaded. This allowed me to have the desired tab immediately focused in the Filter Loaded view because your "filter keep selected" option is now activated.

**Reply by madprops** [reply]
Ok I'm modifying step back to have a step where it selects active tab if not selected, while keeping the filter.
